akka-stream/src/main/scala/akka/stream/impl/io/SslTls.scala

```diff
@@ -0,0 +1,419 @@
+    if (continue && outboundHalfClosed.isReady) {
+      if (tracing) log.debug("outboundClosed continue")
+      try doWrap()
+      catch { case ex: SSLException ⇒ nextPhase(completedPhase) }
+    }
+  }
+
+  val inboundClosed = TransferPhase(outbound || inboundHalfClosed) { () ⇒
+    if (tracing) log.debug("inboundClosed")
+    val continue = doInbound(isOutboundClosed = false, inboundHalfClosed)
+    if (continue) {
+      if (tracing) log.debug("inboundClosed continue")
+      doOutbound(isInboundClosed = true)
+    }
+  }
+
+  def flushToTransport(): Unit = {
+    if (tracing) log.debug("flushToTransport")
+    transportOutBuffer.flip()
+    if (transportOutBuffer.hasRemaining) {
+      val bs = ByteString(transportOutBuffer)
+      outputBunch.enqueue(TransportOut, bs)
+      if (tracing) log.debug(s"sending ${bs.size} bytes")
+    }
+    transportOutBuffer.clear()
+  }
+
+  def flushToUser(): Unit = {
+    if (tracing) log.debug("flushToUser")
+    userOutBuffer.flip()
+    if (userOutBuffer.hasRemaining) {
+      val bs = ByteString(userOutBuffer)
+      outputBunch.enqueue(UserOut, SessionBytes(currentSession, bs))
+    }
+    userOutBuffer.clear()
+  }
+
+  private def doWrap(): Unit = {
+    val result = engine.wrap(userInBuffer, transportOutBuffer)
+    lastHandshakeStatus = result.getHandshakeStatus
+    if (tracing) log.debug(s"wrap: status=${result.getStatus} handshake=$lastHandshakeStatus remaining=${userInBuffer.remaining} out=${transportOutBuffer.position}")
+    if (lastHandshakeStatus == FINISHED) handshakeFinished()
+    runDelegatedTasks()
+    result.getStatus match {
+      case OK ⇒
+        flushToTransport()
+        userInChoppingBlock.putBack(userInBuffer)
+      case CLOSED ⇒
+        flushToTransport()
+        if (engine.isInboundDone()) nextPhase(completedPhase)
+      case s ⇒ fail(new IllegalStateException(s"unexpected status $s in doWrap()"))
+    }
+  }
+
+  @tailrec
+  private def doUnwrap(): Unit = {
+    val result = engine.unwrap(transportInBuffer, userOutBuffer)
+    lastHandshakeStatus = result.getHandshakeStatus
+    if (tracing) log.debug(s"unwrap: status=${result.getStatus} handshake=$lastHandshakeStatus remaining=${transportInBuffer.remaining} out=${userOutBuffer.position}")
+    runDelegatedTasks()
+    result.getStatus match {
+      case OK ⇒
+        result.getHandshakeStatus match {
+          case NEED_WRAP ⇒ flushToUser()
+          case FINISHED ⇒
+            flushToUser()
+            handshakeFinished()
+            transportInChoppingBlock.putBack(transportInBuffer)
+          case _ ⇒
+            if (transportInBuffer.hasRemaining()) doUnwrap()
+            else flushToUser()
+        }
+      case CLOSED ⇒
+        flushToUser()
+        nextPhase(flushingOutbound)
+      case BUFFER_UNDERFLOW ⇒
+        flushToUser()
+      case BUFFER_OVERFLOW ⇒
+        flushToUser()
+        transportInChoppingBlock.putBack(transportInBuffer)
```

Comment:
ah, because we failed the unwrap, we need to get our bytes back into place to retry, right?